src/acl_kernel.cpp

```diff
@@ -2770,6 +2770,9 @@ l_copy_and_adjust_arguments_for_device(cl_kernel kernel, cl_device_id device,
     next_local[aspace.aspace_id] +=
         l_round_up_for_alignment(aspace.static_demand);
   }
+
+  streaming_info.args.clear();
```

Comment:
This segfaults on [Linux](https://github.com/intel/fpga-runtime-for-opencl/runs/6315849715?check_suite_focus=true) and [Windows](https://github.com/intel/fpga-runtime-for-opencl/runs/6315849447?check_suite_focus=true).

Reply:
Manually reproduced backtrace:
```
(gdb) r -v -g kernel
Starting program: /nfs/sc/disks/swuser_work_pcolberg/git/github.com/intel/fpga-runtime-for-opencl/streaming/build/debug/test/acl_test -v -g kernel
Missing separate debuginfos, use: zypper install glibc-debuginfo-2.22-100.15.4.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
warning: File "/nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/lib64/libstdc++.so.6.0.24-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
	add-auto-load-safe-path /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/lib64/libstdc++.so.6.0.24-gdb.py
line to your configuration file "/nfs/site/home/pcolberg/.gdbinit".
To completely disable this security protection add
	set auto-load safe-path /
line to your configuration file "/nfs/site/home/pcolberg/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
	info "(gdb)Auto-loading safe path"
sizeof platform 753160
sizeof context 1648
sizeof command queue 232. Initially allocated 1024 queues
sizeof event 400
sizeof cl_mem 6360
sizeof kernel 248
sizeof device_op_queue 214144
[Detaching after fork from child process 23471]
CreateBuffer
CreateBuffer Finished:  7b30030
[Detaching after fork from child process 23474]
TEST(acl_kernel, set_kernel_exec_info)CreateBuffer
CreateBuffer Finished:  7b31390
CreateBuffer
CreateBuffer Finished:  7b38350
CreateBuffer
CreateBuffer Finished:  7b3ca00
 - 1 ms
TEST(acl_kernel, set_usm_kernel_arg)CreateBuffer
CreateBuffer Finished:  7b31440
CreateBuffer
CreateBuffer Finished:  7b38490
CreateBuffer
CreateBuffer Finished:  7b3caa0
CreateBuffer
CreateBuffer Finished:  7b42d20
acl_do_physical_buffer_allocation
resizing reserved_allocations, physical_device_id:0, target_size:1 
acl_allocate_block size:1000, working_range:50ef400 - 70ef400, initial try:50ef400 
acl_allocate_block finished: result:1, block_allocation:50ef400 - 50f0400 
 - 0 ms
TEST(acl_kernel, set_svm_kernel_arg)CreateBuffer
CreateBuffer Finished:  7b313a0
CreateBuffer
CreateBuffer Finished:  7b38380
CreateBuffer
CreateBuffer Finished:  7b3ca00
 - 0 ms
TEST(acl_kernel_reprogram_scheduler, printf_handler)CreateBuffer
CreateBuffer Finished:  7b2f100
CreateBuffer
resizing reserved_allocations, physical_device_id:0, target_size:1 
CreateBuffer Finished:  7b877b0
CreateBuffer
resizing reserved_allocations, physical_device_id:0, target_size:1 
CreateBuffer Finished:  7b994a0
kernel args
arg 0 mem_obj 7b994a0 regular buffer needed_physical_id 0 needed_mem_id 0 already reserved (reserved allocation 7b4f100 block allocation 7b4f100) count 1 
arg 1 mem_obj 7b877b0 regular buffer needed_physical_id 0 needed_mem_id 0 already reserved (reserved allocation 7b87780 block allocation 7b87780) count 1 
arg 2 const
acl_mem_migrate_buffer
object 0, mem 7b994a0, count 1:already at dest
Memory op is already complete
acl_mem_migrate_buffer finished
acl_mem_migrate_buffer
object 1, mem 7b877b0, count 1:already at dest
Memory op is already complete
acl_mem_migrate_buffer finished
acl_launch_kernel
(null)(null) - 1 ms
TEST(acl_kernel_reprogram_scheduler, use_host_buf_use_twice_same_invocation)CreateBuffer
CreateBuffer Finished:  7b2f100
CreateBuffer
resizing reserved_allocations, physical_device_id:0, target_size:1 
CreateBuffer Finished:  7b879f0

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff69c79f4 in free () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff69c79f4 in free () from /lib64/libc.so.6
#1  0x000000000043df3c in __gnu_cxx::new_allocator<aocl_mmd_streaming_kernel_arg_info_t>::deallocate (
    __p=0x2020202020202020, this=0x7fffffffc900)
    at /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/include/c++/7.4.0/ext/new_allocator.h:125
#2  std::allocator_traits<std::allocator<aocl_mmd_streaming_kernel_arg_info_t> >::deallocate (__n=<optimized out>, 
    __p=0x2020202020202020, __a=...)
    at /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/include/c++/7.4.0/bits/alloc_traits.h:462
#3  std::_Vector_base<aocl_mmd_streaming_kernel_arg_info_t, std::allocator<aocl_mmd_streaming_kernel_arg_info_t> >::_M_deallocate (__n=<optimized out>, __p=0x2020202020202020, this=0x7fffffffc900)
    at /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/include/c++/7.4.0/bits/stl_vector.h:180
#4  std::_Vector_base<aocl_mmd_streaming_kernel_arg_info_t, std::allocator<aocl_mmd_streaming_kernel_arg_info_t> >::~_Vector_base (this=0x7fffffffc900, __in_chrg=<optimized out>)
    at /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/include/c++/7.4.0/bits/stl_vector.h:162
#5  std::vector<aocl_mmd_streaming_kernel_arg_info_t, std::allocator<aocl_mmd_streaming_kernel_arg_info_t> >::~vector
    (this=0x7fffffffc900, __in_chrg=<optimized out>)
    at /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/include/c++/7.4.0/bits/stl_vector.h:435
#6  std::vector<aocl_mmd_streaming_kernel_arg_info_t, std::allocator<aocl_mmd_streaming_kernel_arg_info_t> >::_M_move_assign (__x=..., this=<optimized out>, this=<optimized out>)
    at /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/include/c++/7.4.0/bits/stl_vector.h:1536
#7  std::vector<aocl_mmd_streaming_kernel_arg_info_t, std::allocator<aocl_mmd_streaming_kernel_arg_info_t> >::operator= (__x=..., this=0x7b4f528)
    at /nfs/site/disks/psg_ctools_1/gcc/7.4.0/2/linux64/rhel7/include/c++/7.4.0/bits/stl_vector.h:466
#8  aocl_mmd_streaming_kernel_info_t::operator= (this=0x7b4f528) at ../../include/acl_hal_mmd.h:47
#9  l_copy_and_adjust_arguments_for_device (kernel=kernel@entry=0x7b4efe0, 
    device=device@entry=0x1aa02d8 <acl_platform+120>, buf=0x7b4f550 "", num_bytes=num_bytes@entry=0x7fffffffc9ec, 
    memory_migration=memory_migration@entry=0x7fffffffca70, streaming_info=...) at ../../src/acl_kernel.cpp:2777
#10 0x00000000004403ef in l_enqueue_kernel_with_type (command_queue=command_queue@entry=0x7b69350, 
    kernel=kernel@entry=0x7b4efe0, work_dim=work_dim@entry=1, global_work_offset=global_work_offset@entry=0x0, 
    global_work_size=global_work_size@entry=0x7fffffffcc40, local_work_size=local_work_size@entry=0x7fffffffcc48, 
    num_events_in_wait_list=0, event_wait_list=0x0, event=0x7fffffffcca0, type=4593) at ../../src/acl_kernel.cpp:2243
#11 0x0000000000440caa in clEnqueueTaskIntelFPGA (command_queue=0x7b69350, kernel=kernel@entry=0x7b4efe0, 
    num_events_in_wait_list=num_events_in_wait_list@entry=0, event_wait_list=event_wait_list@entry=0x0, 
    event=event@entry=0x7fffffffcca0) at ../../src/acl_kernel.cpp:1525
#12 0x0000000000440ccd in clEnqueueTask (command_queue=<optimized out>, kernel=kernel@entry=0x7b4efe0, 
    num_events_in_wait_list=num_events_in_wait_list@entry=0, event_wait_list=event_wait_list@entry=0x0, 
    event=event@entry=0x7fffffffcca0) at ../../src/acl_kernel.cpp:1541
#13 0x00000000005e7d1e in acl_kernel_reprogram_scheduler_use_host_buf_use_twice_same_invocation_Test::testBody (
    this=0x7b9d6b0) at ../../test/acl_kernel_test.cpp:4392
#14 0x00000000006eda41 in Utest::executePlatformSpecificTestBody (this=0x7b9d6b0)
    at ../../lib/CppUTest/src/UtestPlatformGcc.cpp:82
#15 0x00000000006ec65b in Utest::runInThread (this=this@entry=0x7b9d6b0) at ../../lib/CppUTest/src/Utest.cpp:99
#16 0x0000000000627e38 in acl_kernel_reprogram_scheduler_use_host_buf_use_twice_same_invocation_Test::runInThreadOnCopy (this=0x7123300 <acl_kernel_reprogram_scheduler_use_host_buf_use_twice_same_invocation_Instance>)
    at ../../test/acl_kernel_test.cpp:4369
#17 0x00000000006edc8c in Utest::executePlatformSpecificRunInThreads (this=<optimized out>)
    at ../../lib/CppUTest/src/UtestPlatformGcc.cpp:175
#18 0x00000000006ec617 in Utest::run (
    this=0x7123300 <acl_kernel_reprogram_scheduler_use_host_buf_use_twice_same_invocation_Instance>, result=...)
    at ../../lib/CppUTest/src/Utest.cpp:83
#19 0x00000000006ec5c2 in Utest::runOneTest (
    this=0x7123300 <acl_kernel_reprogram_scheduler_use_host_buf_use_twice_same_invocation_Instance>, 
    plugin=0x7fffffffcee0, result=...) at ../../lib/CppUTest/src/Utest.cpp:69
#20 0x00000000006edb18 in Utest::executePlatformSpecificRunOneTest (
    this=0x7123300 <acl_kernel_reprogram_scheduler_use_host_buf_use_twice_same_invocation_Instance>, 
    plugin=0x7fffffffcee0, result=...) at ../../lib/CppUTest/src/UtestPlatformGcc.cpp:100
#21 0x00000000006ec597 in Utest::runOneTestWithPlugins (this=<optimized out>, plugin=<optimized out>, result=...)
    at ../../lib/CppUTest/src/Utest.cpp:63
#22 0x00000000006ec066 in TestRegistry::runAllTests (this=0x7b0bac0 <TestRegistry::getCurrentRegistry()::registry>, 
    result=...) at ../../lib/CppUTest/src/TestRegistry.cpp:72
#23 0x00000000006e91a3 in CommandLineTestRunner::runAllTests (this=this@entry=0x7fffffffcf30)
    at ../../lib/CppUTest/src/CommandLineTestRunner.cpp:105
#24 0x00000000006e9266 in CommandLineTestRunner::runAllTestsMain (this=this@entry=0x7fffffffcf30)
    at ../../lib/CppUTest/src/CommandLineTestRunner.cpp:81
#25 0x00000000006e9332 in CommandLineTestRunner::RunAllTests (ac=4, av=0x7fffffffd0e8)
    at ../../lib/CppUTest/src/CommandLineTestRunner.cpp:63
#26 0x00000000006d6c02 in main (argc=4, argv=0x7fffffffd0e8) at ../../test/acl_test.cpp:90
(gdb) 
```

Reply:
This might be related to the way in which the pool of kernel invocation wrappers is allocated:
https://github.com/intel/fpga-runtime-for-opencl/blob/bd7f1073b7a85858b5180a3329a01286f34bc8d8/src/acl_context.cpp#L1219-L1296

Reply:
The segmentation fault is indeed resolved by constructing the streaming object in-place:
```diff
--- a/src/acl_kernel.cpp
+++ b/src/acl_kernel.cpp
@@ -2147,6 +2147,7 @@ static cl_int l_enqueue_kernel_with_type(
                 "Could not allocate a kernel invocation object");
       }
     }
+    new (&serialization_wrapper->streaming_info) aocl_mmd_streaming_kernel_info_t{};
     acl_retain(serialization_wrapper);
     serialization_wrapper->event = 0;
 
@@ -2198,6 +2199,7 @@ static cl_int l_enqueue_kernel_with_type(
               "Could not allocate a kernel invocation object");
     }
   }
+  new (&wrapper->streaming_info) aocl_mmd_streaming_kernel_info_t{};
 
   acl_retain(wrapper);
   wrapper->event = 0;
```
However, the above still lacks explicit calls to the destructor and thus leak memory allocated for the arguments.

Reply:
This is addressed in commit https://github.com/intel/fpga-runtime-for-opencl/pull/103/commits/7a29c63d717b723d80017615e3f4dba994629c49. I am not yet sure why placement-new of `wrapper->streaming_info` works, while placement-new of the whole `wrapper` still produces a segmentation fault.

Reply:
Never mind, placement-new of the whole `wrapper` has to be done **before** assigning to any of its members, otherwise those will be reinitialized with the default values, see commit https://github.com/intel/fpga-runtime-for-opencl/pull/103/commits/bf037d20c8df924e499423cf3e683d59cfa5cb45 :facepalm:

Reply:
Resolved in https://github.com/intel/fpga-runtime-for-opencl/pull/115